File: apps/clawminer/internal/relay/relay.go
// Package relay implements the SPV Relay Mesh for ClawMiner.
//
// It provides an in-memory transaction cache and HTTP endpoints so peers can
// fetch recently broadcast transactions even if they missed the GossipSub message.
package relay

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"regexp"
	"sync"
	"time"
)

const (
	maxCacheSize  = 10_000
	cacheTTL      = 1 * time.Hour
	pruneInterval = 60 * time.Second
)

var txidRe = regexp.MustCompile(`^[a-fA-F0-9]{64}$`)

// CachedTx holds a recently seen transaction.
type CachedTx struct {
	RawHex    string `json:"raw_hex"`
	Confirmed bool   `json:"confirmed"`
	BlockHash string `json:"block_hash,omitempty"`
	Source    string `json:"source,omitempty"`
	StoredAt  time.Time
}

// Service is the relay mesh cache and HTTP server.
type Service struct {
	mu        sync.RWMutex
	cache     map[string]*CachedTx
	startTime time.Time
	peerCount func() int // callback to get current peer count
	stopCh    chan struct{}
}

// New creates a relay service. peerCountFn should return the current gossip peer count.
func New(peerCountFn func() int) *Service {
	return &Service{
		cache:     make(map[string]*CachedTx),
		peerCount: peerCountFn,
		stopCh:    make(chan struct{}),
	}
}

// Start begins the background pruning goroutine.
func (s *Service) Start() {
	s.startTime = time.Now()
	go s.pruneLoop()
	log.Println("[relay] Service started")
}

// Stop halts the background pruning.
func (s *Service) Stop() {
	close(s.stopCh)
	log.Println("[relay] Service stopped")
}

// StoreTx stores a transaction in the in-memory cache.
func (s *Service) StoreTx(txid, rawHex string, confirmed bool, blockHash, source string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Evict oldest if at capacity
	if len(s.cache) >= maxCacheSize {
		s.evictOldest()
	}

	if existing, ok := s.cache[txid]; ok {
		// Update confirmation status if confirmed
		if confirmed && !existing.Confirmed {
			existing.Confirmed = true
			existing.BlockHash = blockHash
		}
		return
	}

	s.cache[txid] = &CachedTx{
		RawHex:    rawHex,
		Confirmed: confirmed,
		BlockHash: blockHash,
		Source:    source,
		StoredAt:  time.Now(),
	}
}

// GetTx retrieves a transaction from the cache.
func (s *Service) GetTx(txid string) *CachedTx {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.cache[txid]
}

// HasTx checks if a transaction exists in the cache.
func (s *Service) HasTx(txid string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	_, ok := s.cache[txid]
	return ok
}

// CacheSize returns the number of transactions in the cache.
func (s *Service) CacheSize() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.cache)
}

func (s *Service) evictOldest() {
	var oldestKey string
	var oldestTime time.Time
	first := true
	for k, v := range s.cache {
		if first || v.StoredAt.Before(oldestTime) {
			oldestKey = k
			oldestTime = v.StoredAt
			first = false
		}
	}
	if oldestKey != "" {
		delete(s.cache, oldestKey)
	}
}

func (s *Service) pruneLoop() {
	ticker := time.NewTicker(pruneInterval)
	defer ticker.Stop()
	for {
		select {
		case <-s.stopCh:
			return
		case <-ticker.C:
			s.prune()
		}
	}
}

func (s *Service) prune() {
	s.mu.Lock()
	defer s.mu.Unlock()
	cutoff := time.Now().Add(-cacheTTL)
	pruned := 0
	for k, v := range s.cache {
		if v.StoredAt.Before(cutoff) {
			delete(s.cache, k)
			pruned++
		}
	}
	if pruned > 0 {
		log.Printf("[relay] Pruned %d expired entries, %d remaining", pruned, len(s.cache))
	}
}

// ── HTTP Handlers ─────────────────────────────────────────────────

// RegisterRoutes mounts relay endpoints on the provided mux.
func (s *Service) RegisterRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /relay/tx/{txid}", s.handleGetTx)
	mux.HandleFunc("GET /relay/tx/{txid}/status", s.handleGetTxStatus)
	mux.HandleFunc("GET /relay/health", s.handleHealth)
	mux.HandleFunc("POST /relay/tx", s.handlePostTx)
}

func (s *Service) handleGetTx(w http.ResponseWriter, r *http.Request) {
	txid := r.PathValue("txid")
	if txid == "" || !txidRe.MatchString(txid) {
		writeRelayError(w, 400, "invalid txid")
		return
	}

	tx := s.GetTx(txid)
	if tx == nil {
		writeRelayError(w, 404, "tx not found")
		return
	}

	writeRelayJSON(w, map[string]interface{}{
		"txid":       txid,
		"raw_hex":    tx.RawHex,
		"confirmed":  tx.Confirmed,
		"block_hash": tx.BlockHash,
	})
}

func (s *Service) handleGetTxStatus(w http.ResponseWriter, r *http.Request) {
	txid := r.PathValue("txid")
	if txid == "" || !txidRe.MatchString(txid) {
		writeRelayError(w, 400, "invalid txid")
		return
	}

	tx := s.GetTx(txid)
	writeRelayJSON(w, map[string]interface{}{
		"txid":       txid,
		"found":      tx != nil,
		"confirmed":  tx != nil && tx.Confirmed,
		"block_hash": func() string { if tx != nil { return tx.BlockHash }; return "" }(),
	})
}

func (s *Service) handleHealth(w http.ResponseWriter, r *http.Request) {
	peers := 0
	if s.peerCount != nil {
		peers = s.peerCount()
	}
	writeRelayJSON(w, map[string]interface{}{
		"peer_count": peers,
		"cache_size": s.CacheSize(),
		"uptime_ms":  time.Since(s.startTime).Milliseconds(),
	})
}

func (s *Service) handlePostTx(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Txid   string `json:"txid"`
		RawHex string `json:"raw_hex"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeRelayError(w, 400, fmt.Sprintf("invalid json: %v", err))
		return
	}

	if body.Txid == "" || body.RawHex == "" {
		writeRelayError(w, 400, "txid and raw_hex required")
		return
	}

	if !txidRe.MatchString(body.Txid) {
		writeRelayError(w, 400, "invalid txid format")
		return
	}

	nomesh := r.URL.Query().Get("nomesh") == "1"
	s.StoreTx(body.Txid, body.RawHex, false, "", "http")

	writeRelayJSON(w, map[string]interface{}{
		"stored":  true,
		"relayed": !nomesh,
	})
}

func writeRelayJSON(w http.ResponseWriter, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v)
}

func writeRelayError(w http.ResponseWriter, code int, msg string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	json.NewEncoder(w).Encode(map[string]string{"error": msg})
}

File: packages/core/src/services/relay.ts
/**
 * SPV Relay Mesh — RelayService
 *
 * When a ClawMiner broadcasts a mint tx, peers get it instantly via gossip.
 * If gossip misses a tx, peers request it over HTTP (mesh fallback).
 *
 * Memory cache (fast, bounded) + SQLite (persistent, survives restart).
 * HTTP handlers plug into the agent's existing Express/Fastify server.
 */

import { EventEmitter } from 'events';
import type { GossipNode } from '../gossip/node.js';
import {
  upsertRelayTx,
  getRelayTx,
  hasRelayTx,
  getRelayTxCount,
  pruneRelayTxs,
} from '../db/index.js';

// ── Constants ─────────────────────────────────────────────────────

const MAX_CACHE_SIZE = 10_000;
const CACHE_TTL_MS = 60 * 60 * 1000; // 1 hour
const PRUNE_INTERVAL_MS = 60 * 1000;  // 60 seconds
const MESH_TIMEOUT_MS = 5_000;

// ── Types ─────────────────────────────────────────────────────────

interface CachedTx {
  hex: string;
  ts: number;
  confirmed: boolean;
  blockHash?: string;
}

export interface RelayHealthInfo {
  peer_count: number;
  cache_size: number;
  db_size: number;
  uptime_ms: number;
}

// ── RelayService ──────────────────────────────────────────────────

export class RelayService extends EventEmitter {
  private txCache: Map<string, CachedTx> = new Map();
  private gossipNode: GossipNode | null = null;
  private pruneTimer: ReturnType<typeof setInterval> | null = null;
  private startTime: number = Date.now();
  private meshPeers: Set<string> = new Set();

  constructor() {
    super();
  }

  // ── Cache Operations ──────────────────────────────────────────

  storeTx(txid: string, hex: string, confirmed = false, blockHash?: string, sourcePeer?: string): void {
    // Memory cache
    if (this.txCache.size >= MAX_CACHE_SIZE) {
      // Evict oldest entry
      const oldest = this.txCache.entries().next().value;
      if (oldest) this.txCache.delete(oldest[0]);
    }

    this.txCache.set(txid, {
      hex,
      ts: Date.now(),
      confirmed,
      blockHash,
    });

    // Persistent store
    try {
      upsertRelayTx({ txid, raw_hex: hex, confirmed, block_hash: blockHash, source_peer: sourcePeer });
    } catch {
      // DB not initialized in some environments
    }

    this.emit('tx:stored', txid);
  }

  getTx(txid: string): { hex: string; confirmed: boolean; blockHash?: string } | null {
    // Check memory first
    const cached = this.txCache.get(txid);
    if (cached) {
      return { hex: cached.hex, confirmed: cached.confirmed, blockHash: cached.blockHash };
    }

    // Fall back to DB
    try {
      const row = getRelayTx(txid);
      if (row) {
        // Warm the memory cache
        this.txCache.set(txid, {
          hex: row.raw_hex,
          ts: row.created_at * 1000,
          confirmed: row.confirmed === 1,
          blockHash: row.block_hash ?? undefined,
        });
        return { hex: row.raw_hex, confirmed: row.confirmed === 1, blockHash: row.block_hash ?? undefined };
      }
    } catch {
      // DB not available
    }

    return null;
  }

  hasTx(txid: string): boolean {
    if (this.txCache.has(txid)) return true;
    try { return hasRelayTx(txid); } catch { return false; }
  }

  getCacheSize(): number {
    return this.txCache.size;
  }

  // ── Gossip Integration ────────────────────────────────────────

  attachToGossip(node: GossipNode): void {
    this.gossipNode = node;

    // When a peer relays a tx via gossip, store it locally
    node.on('tx:relayed', (txid: string, rawHex: string, peerId: string) => {
      if (!this.hasTx(txid)) {
        this.storeTx(txid, rawHex, false, undefined, peerId);
        console.log(`[Relay] Stored relayed TX ${txid.slice(0, 16)}... from peer ${peerId.slice(0, 8)}...`);
      }
    });

    // When a peer requests a tx, check our cache
    node.on('tx:requested', (txid: string, peerId: string) => {
      const tx = this.getTx(txid);
      if (tx) {
        console.log(`[Relay] Serving TX ${txid.slice(0, 16)}... to peer ${peerId.slice(0, 8)}...`);
        this.emit('tx:served', txid, peerId);
      }
    });
  }

  /**
   * Called after a successful mint broadcast.
   * Stores the tx locally and relays to all peers via gossip.
   */
  onMintBroadcast(txid: string, rawHex: string): void {
    this.storeTx(txid, rawHex, false);

    if (this.gossipNode) {
      this.gossipNode.relayTransaction(txid, rawHex);
    }

    console.log(`[Relay] Mint TX ${txid.slice(0, 16)}... stored and relayed`);
    this.emit('tx:minted', txid);
  }

  // ── Mesh Fallback (HTTP) ──────────────────────────────────────

  /**
   * Register a mesh peer URL for HTTP fallback lookups.
   */
  addMeshPeer(url: string): void {
    this.meshPeers.add(url);
  }

  removeMeshPeer(url: string): void {
    this.meshPeers.delete(url);
  }

  /**
   * Look up a tx from mesh peers via HTTP.
   * Tries each peer until one responds with the tx.
   */
  async meshLookup(txid: string, excludePeer?: string): Promise<string | null> {
    for (const peerUrl of this.meshPeers) {
      if (excludePeer && peerUrl === excludePeer) continue;

      try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), MESH_TIMEOUT_MS);

        const res = await fetch(`${peerUrl}/relay/tx/${txid}?nomesh=1`, {
          signal: controller.signal,
        });
        clearTimeout(timeout);

        if (res.ok) {
          const data = await res.json() as { txid: string; raw_hex: string; confirmed: boolean; block_hash?: string };
          if (data.raw_hex) {
            this.storeTx(txid, data.raw_hex, data.confirmed, data.block_hash, peerUrl);
            console.log(`[Relay] Mesh fetched TX ${txid.slice(0, 16)}... from ${peerUrl}`);
            return data.raw_hex;
          }
        }
      } catch {
        // Peer unreachable — continue to next
      }
    }

    return null;
  }

  // ── Lifecycle ─────────────────────────────────────────────────

  start(): void {
    this.startTime = Date.now();

    // Prune expired entries periodically
    this.pruneTimer = setInterval(() => {
      const now = Date.now();
      let pruned = 0;
      for (const [txid, entry] of this.txCache) {
        if (now - entry.ts > CACHE_TTL_MS) {
          this.txCache.delete(txid);
          pruned++;
        }
      }
      // Also prune DB
      try { pruneRelayTxs(CACHE_TTL_MS / 1000); } catch { /* no db */ }
      if (pruned > 0) {
        console.log(`[Relay] Pruned ${pruned} expired entries from memory cache`);
      }
    }, PRUNE_INTERVAL_MS);

    console.log('[Relay] RelayService started');
  }

  stop(): void {
    if (this.pruneTimer) {
      clearInterval(this.pruneTimer);
      this.pruneTimer = null;
    }
    console.log('[Relay] RelayService stopped');
  }

  health(): RelayHealthInfo {
    let dbSize = 0;
    try { dbSize = getRelayTxCount(); } catch { /* no db */ }

    return {
      peer_count: this.gossipNode?.getPeerCount() ?? 0,
      cache_size: this.txCache.size,
      db_size: dbSize,
      uptime_ms: Date.now() - this.startTime,
    };
  }

  // ── HTTP Route Handlers ───────────────────────────────────────
  // These return handler functions suitable for Express, Fastify, or raw http

  /**
   * GET /relay/tx/:txid — return raw tx hex from cache
   * Query: ?nomesh=1 disables mesh forwarding (loop prevention)
   */
  handleGetTx() {
    return async (txid: string, nomesh: boolean): Promise<{
      status: number;
      body: Record<string, unknown>;
    }> => {
      const tx = this.getTx(txid);
      if (tx) {
        return {
          status: 200,
          body: { txid, raw_hex: tx.hex, confirmed: tx.confirmed, block_hash: tx.blockHash ?? null },
        };
      }

      // Mesh fallback (unless nomesh)
      if (!nomesh) {
        const hex = await this.meshLookup(txid);
        if (hex) {
          const stored = this.getTx(txid);
          return {
            status: 200,
            body: { txid, raw_hex: hex, confirmed: stored?.confirmed ?? false, block_hash: stored?.blockHash ?? null },
          };
        }
      }

      return { status: 404, body: { error: 'tx not found' } };
    };
  }

  /**
   * GET /relay/tx/:txid/status — check if we have a tx
   */
  handleGetTxStatus() {
    return (txid: string): { status: number; body: Record<string, unknown> } => {
      const tx = this.getTx(txid);
      return {
        status: 200,
        body: {
          txid,
          found: !!tx,
          confirmed: tx?.confirmed ?? false,
          block_hash: tx?.blockHash ?? null,
        },
      };
    };
  }

  /**
   * GET /relay/health
   */
  handleHealth() {
    return (): { status: number; body: RelayHealthInfo } => {
      return { status: 200, body: this.health() };
    };
  }

  /**
   * POST /relay/tx — accept raw tx, store + relay
   * Body: { txid: string, raw_hex: string }
   * Query: ?nomesh=1 disables mesh forwarding
   */
  handlePostTx() {
    return (txid: string, rawHex: string, nomesh: boolean): { status: number; body: Record<string, unknown> } => {
      if (!txid || !rawHex) {
        return { status: 400, body: { error: 'txid and raw_hex required' } };
      }

      // Basic txid format validation
      if (!/^[a-fA-F0-9]{64}$/.test(txid)) {
        return { status: 400, body: { error: 'invalid txid format' } };
      }

      this.storeTx(txid, rawHex);

      // Relay to gossip (unless nomesh)
      if (!nomesh && this.gossipNode) {
        this.gossipNode.relayTransaction(txid, rawHex);
      }

      return { status: 200, body: { stored: true, relayed: !nomesh } };
    };
  }
}
